File: app/pipelines/expressions/evaluator.py
"""Evaluator for pipeline expressions.

Evaluation is defensive rather than trusting: it re-derives value types at
runtime (environments are built from caller-supplied argument values), so a
value that slipped past static checking still fails with a typed
`ExpressionTypeError` instead of leaking Python semantics like `bool + int`.
Arithmetic failures that types cannot catch (divide by zero, inverted clamp
range) raise `ExpressionEvalError`.
"""

from __future__ import annotations

from collections.abc import Mapping

from app.pipelines.expressions.errors import ExpressionEvalError, ExpressionTypeError
from app.pipelines.expressions.functions import BUILTINS, Numeric, arity_message
from app.pipelines.expressions.parser import (
    Binary,
    BooleanLiteral,
    Call,
    Expression,
    IntLiteral,
    Member,
    Name,
    NumberLiteral,
    StringLiteral,
    Unary,
)
from app.pipelines.expressions.values import (
    MODEL_MEMBERS,
    ExprValue,
    ModelValue,
    value_type,
)


def evaluate(expr: Expression, env: Mapping[str, ExprValue]) -> ExprValue:
    """Evaluate the expression against `{variable name: value}`."""
    if isinstance(expr, (IntLiteral, NumberLiteral, StringLiteral, BooleanLiteral)):
        return expr.value
    if isinstance(expr, Name):
        if expr.name not in env:
            raise ExpressionTypeError(f"Unknown variable '{expr.name}'", expr.position)
        return env[expr.name]
    if isinstance(expr, Member):
        return _evaluate_member(expr, env)
    if isinstance(expr, Unary):
        operand = _require_numeric(evaluate(expr.operand, env), "Unary '-'", expr.position)
        return -operand
    if isinstance(expr, Binary):
        return _evaluate_binary(expr, env)
    if isinstance(expr, Call):
        return _evaluate_call(expr, env)
    raise ExpressionTypeError("Unsupported expression node", expr.position)


def _require_numeric(value: ExprValue, context: str, position: int) -> Numeric:
    """Narrow a value to int/float or raise a typed error.

    Booleans are excluded explicitly: Python's `bool` subclasses `int`, and
    `flag * 2` must fail here exactly as it does in static checking.
    """
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ExpressionTypeError(
            f"{context} requires a number, got {value_type(value)}", position
        )
    return value


def _require_integer(value: ExprValue, op: str, position: int) -> int:
    """Narrow a value to a non-boolean int or raise a typed error."""
    if isinstance(value, bool) or not isinstance(value, int):
        raise ExpressionTypeError(
            f"'{op}' requires integers, got {value_type(value)}", position
        )
    return value


def _evaluate_member(expr: Member, env: Mapping[str, ExprValue]) -> ExprValue:
    """Evaluate model member access to its string value."""
    base = evaluate(expr.base, env)
    if not isinstance(base, ModelValue) or expr.attribute not in MODEL_MEMBERS:
        raise ExpressionTypeError(
            f"Cannot access '{expr.attribute}' on {value_type(base)}", expr.position
        )
    if expr.attribute == "connection_id":
        return str(base.connection_id)
    return base.model_name


def _evaluate_binary(expr: Binary, env: Mapping[str, ExprValue]) -> ExprValue:
    """Evaluate a binary operation, mirroring `_check_binary`'s rules."""
    left = evaluate(expr.left, env)
    right = evaluate(expr.right, env)
    if expr.op == "+" and isinstance(left, str) and isinstance(right, str):
        return left + right
    if expr.op in ("//", "%"):
        left_int = _require_integer(left, expr.op, expr.position)
        right_int = _require_integer(right, expr.op, expr.position)
        if right_int == 0:
            raise ExpressionEvalError(f"'{expr.op}' by zero", expr.position)
        return left_int // right_int if expr.op == "//" else left_int % right_int
    left_num = _require_numeric(left, f"'{expr.op}'", expr.position)
    right_num = _require_numeric(right, f"'{expr.op}'", expr.position)
    if expr.op == "+":
        return left_num + right_num
    if expr.op == "-":
        return left_num - right_num
    if expr.op == "*":
        return left_num * right_num
    if right_num == 0:
        raise ExpressionEvalError("'/' by zero", expr.position)
    return left_num / right_num


def _evaluate_call(expr: Call, env: Mapping[str, ExprValue]) -> ExprValue:
    """Evaluate a builtin call against the shared catalog."""
    spec = BUILTINS.get(expr.name)
    if spec is None:
        raise ExpressionTypeError(f"Unknown function '{expr.name}'", expr.position)
    received = len(expr.args)
    if received < spec.min_args or (spec.max_args is not None and received > spec.max_args):
        raise ExpressionTypeError(arity_message(spec, received), expr.position)
    args = [
        _require_numeric(evaluate(arg, env), f"{spec.name}()", arg.position)
        for arg in expr.args
    ]
    try:
        return spec.apply(args)
    except ExpressionEvalError as error:
        raise ExpressionEvalError(error.message, expr.position) from error

File: frontend/src/lib/expressions/evaluator.ts
/**
 * Evaluator — the TypeScript mirror of `app/pipelines/expressions/evaluator.py`.
 * Floor division and modulo follow floor semantics (Python's), pinned by the
 * shared vectors: `-7 // 2 === -4`, `-7 % 3 === 2`.
 */

import { evalError, typeError } from "./errors";
import { BUILTINS, arityMessage } from "./functions";
import { MODEL_MEMBERS, isModelValue, valueType, type ExprValue } from "./values";

import type { Expression } from "./parser";

export type ValueEnvironment = ReadonlyMap<string, ExprValue>;

export function evaluate(expr: Expression, env: ValueEnvironment): ExprValue {
  switch (expr.kind) {
    case "int":
    case "float":
    case "string":
    case "boolean":
      return expr.value;
    case "name": {
      const value = env.get(expr.name);
      if (value === undefined) {
        throw typeError(`Unknown variable '${expr.name}'`, expr.position);
      }
      return value;
    }
    case "member":
      return evaluateMember(expr, env);
    case "unary":
      return -requireNumeric(evaluate(expr.operand, env), "Unary '-'", expr.position);
    case "binary":
      return evaluateBinary(expr, env);
    case "call":
      return evaluateCall(expr, env);
  }
}

function requireNumeric(value: ExprValue, context: string, position: number): number {
  if (typeof value !== "number") {
    throw typeError(`${context} requires a number, got ${valueType(value)}`, position);
  }
  return value;
}

function requireInteger(value: ExprValue, op: string, position: number): number {
  if (typeof value !== "number" || !Number.isInteger(value)) {
    throw typeError(`'${op}' requires integers, got ${valueType(value)}`, position);
  }
  return value;
}

function evaluateMember(
  expr: Extract<Expression, { kind: "member" }>,
  env: ValueEnvironment,
): ExprValue {
  const base = evaluate(expr.base, env);
  if (!isModelValue(base) || !(expr.attribute in MODEL_MEMBERS)) {
    throw typeError(`Cannot access '${expr.attribute}' on ${valueType(base)}`, expr.position);
  }
  return expr.attribute === "connection_id" ? base.connection_id : base.model_name;
}

function evaluateBinary(
  expr: Extract<Expression, { kind: "binary" }>,
  env: ValueEnvironment,
): ExprValue {
  const left = evaluate(expr.left, env);
  const right = evaluate(expr.right, env);
  if (expr.op === "+" && typeof left === "string" && typeof right === "string") {
    return left + right;
  }
  if (expr.op === "//" || expr.op === "%") {
    const leftInt = requireInteger(left, expr.op, expr.position);
    const rightInt = requireInteger(right, expr.op, expr.position);
    if (rightInt === 0) {
      throw evalError(`'${expr.op}' by zero`, expr.position);
    }
    const quotient = Math.floor(leftInt / rightInt);
    return expr.op === "//" ? quotient : leftInt - quotient * rightInt;
  }
  const leftNum = requireNumeric(left, `'${expr.op}'`, expr.position);
  const rightNum = requireNumeric(right, `'${expr.op}'`, expr.position);
  switch (expr.op) {
    case "+":
      return leftNum + rightNum;
    case "-":
      return leftNum - rightNum;
    case "*":
      return leftNum * rightNum;
    default:
      if (rightNum === 0) {
        throw evalError("'/' by zero", expr.position);
      }
      return leftNum / rightNum;
  }
}

function evaluateCall(
  expr: Extract<Expression, { kind: "call" }>,
  env: ValueEnvironment,
): ExprValue {
  const spec = BUILTINS[expr.name];
  if (!spec) {
    throw typeError(`Unknown function '${expr.name}'`, expr.position);
  }
  const received = expr.args.length;
  if (received < spec.minArgs || (spec.maxArgs !== null && received > spec.maxArgs)) {
    throw typeError(arityMessage(spec, received), expr.position);
  }
  const args = expr.args.map((arg) =>
    requireNumeric(evaluate(arg, env), `${spec.name}()`, arg.position),
  );
  try {
    return spec.apply(args);
  } catch (error) {
    if (error instanceof Error && "kind" in error) {
      throw evalError(error.message, expr.position);
    }
    throw error;
  }
}
